File: src/papyrus_api/routers/cards.py
"""卡片管理API路由。"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from papyrus.core import cards as card_core
from papyrus.core.cards import CardDict
from papyrus_api.deps import get_data_file, pick_card_text
from papyrus.paths import DATABASE_FILE
from papyrus.data.progress import record_card_created

router = APIRouter(prefix="/cards", tags=["cards"])


class CreateCardIn(BaseModel):
    q: str | None = None
    a: str | None = None
    question: str | None = None
    answer: str | None = None
    tags: list[str] | None = None


class CardsListResponse(BaseModel):
    success: bool
    cards: list[CardDict]
    count: int


class CreateCardResponse(BaseModel):
    success: bool
    card: CardDict


class DeleteCardResponse(BaseModel):
    success: bool


class ImportTxtIn(BaseModel):
    content: str


class ImportTxtResponse(BaseModel):
    success: bool
    count: int


@router.get("", response_model=CardsListResponse)
def list_cards() -> CardsListResponse:
    """获取所有卡片列表。"""
    data_file = get_data_file()
    cards = card_core.list_cards(data_file)
    return CardsListResponse(success=True, cards=cards, count=len(cards))


@router.post("", response_model=CreateCardResponse)
def create_card(payload: CreateCardIn) -> CreateCardResponse:
    """创建新卡片。"""
    q = pick_card_text(payload.q, payload.question)
    a = pick_card_text(payload.a, payload.answer)

    data_file = get_data_file()
    try:
        card = card_core.create_card(data_file, q=q, a=a, tags=payload.tags or [])
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    # 记录创建进度
    record_card_created(DATABASE_FILE)

    return CreateCardResponse(success=True, card=card)


@router.delete("/{card_id}", response_model=DeleteCardResponse)
def delete_card(card_id: str) -> DeleteCardResponse:
    """删除指定卡片。"""
    data_file = get_data_file()
    ok = card_core.delete_card(data_file, card_id)
    if not ok:
        raise HTTPException(status_code=404, detail="card not found")
    return DeleteCardResponse(success=True)


class UpdateCardIn(BaseModel):
    q: str | None = None
    a: str | None = None
    tags: list[str] | None = None

class UpdateCardResponse(BaseModel):
    success: bool
    card: CardDict

@router.patch("/{card_id}", response_model=UpdateCardResponse)
def update_card_route(card_id: str, payload: UpdateCardIn) -> UpdateCardResponse:
    """Update a card."""
    data_file = get_data_file()
    card = card_core.update_card(
        data_file,
        card_id,
        q=payload.q,
        a=payload.a,
        tags=payload.tags,
    )
    if card is None:
        raise HTTPException(status_code=404, detail="card not found")
    return UpdateCardResponse(success=True, card=card)


@router.post("/import/txt", response_model=ImportTxtResponse)
def import_txt(payload: ImportTxtIn) -> ImportTxtResponse:
    """从文本导入卡片（格式：问题===答案）。"""
    data_file = get_data_file()
    count = card_core.import_from_txt(data_file, payload.content)
    if count == 0:
        raise HTTPException(status_code=400, detail="未找到有效卡片，请确认格式为：题目===答案")
    return ImportTxtResponse(success=True, count=count)

File: backend/src/api/routes/cards.ts
import type { FastifyInstance } from 'fastify';
import { getAllCards, createCard, updateCard, deleteCard, importCardsFromTxt } from '../../core/cards.js';
import { recordCardCreated } from './progress.js';

export default async function cardsRoutes(fastify: FastifyInstance): Promise<void> {
  fastify.get('/', async (_request, reply) => {
    const cards = getAllCards();
    reply.send({ success: true, cards, count: cards.length });
  });

  fastify.post('/', async (request, reply) => {
    const body = request.body as { q?: string; question?: string; a?: string; answer?: string; tags?: string[] };
    const q = body.q ?? body.question ?? '';
    const a = body.a ?? body.answer ?? '';
    if (!q || !a) {
      reply.status(400).send({ success: false, error: 'Question and answer are required' });
      return;
    }
    const card = createCard(q, a, body.tags ?? []);
    recordCardCreated();
    reply.send({ success: true, card });
  });

  fastify.delete('/:cardId', async (request, reply) => {
    const { cardId } = request.params as { cardId: string };
    const success = deleteCard(cardId);
    if (!success) {
      reply.status(404).send({ success: false, error: 'Card not found' });
      return;
    }
    reply.send({ success: true });
  });

  fastify.patch('/:cardId', async (request, reply) => {
    const { cardId } = request.params as { cardId: string };
    const body = request.body as { q?: string; a?: string; tags?: string[] };
    const card = await updateCard(cardId, body);
    if (!card) {
      reply.status(404).send({ success: false, error: 'Card not found' });
      return;
    }
    reply.send({ success: true, card });
  });

  fastify.post('/import/txt', async (request, reply) => {
    const body = request.body as { content?: string };
    if (!body.content) {
      reply.status(400).send({ success: false, error: 'Content is required' });
      return;
    }
    const cards = importCardsFromTxt(body.content);
    for (const _c of cards) {
      recordCardCreated();
    }
    if (cards.length === 0) {
      reply.status(400).send({ success: false, error: 'No valid cards found' });
      return;
    }
    reply.send({ success: true, count: cards.length });
  });
}
